### ISSUE TITLE
Crash in `-[CASContext JSONPayload]`

### ISSUE THREAD

**Issue by tiwoc** [castle/castle-ios#55]
SDK version: 2.0.0

We're seeing a crash in `-[CASContext JSONPayload]`. We can't reproduce it at will (might be a race condition, see stack trace below), unfortunately.

The crash happens when the `clientID` of a `CASContext` is nil. In this case, line 31 will try to insert nil into a dictionary, which fails with

```
libc++abi.dylib: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSPlaceholderDictionary initWithObjects:forKeys:count:]: attempt to insert nil object from objects[0]'
terminating with uncaught exception of type NSException
```

https://github.com/castle/castle-ios/blob/1066fc026f788ad72197175f2193958e9053414d/Castle/Internal/CASContext.m#L29-L32

`clientID` can be nil because it is loaded from `+[Castle clientId]`, which is nullable:

https://github.com/castle/castle-ios/blob/1066fc026f788ad72197175f2193958e9053414d/Castle/Internal/CASContext.m#L20-L25

https://github.com/castle/castle-ios/blob/1066fc026f788ad72197175f2193958e9053414d/Castle/Public/Castle.h#L172-L177

### Stack Trace
<img width="530" alt="backtrace" src="https://user-images.githubusercontent.com/1260821/116976924-0e640180-acc2-11eb-828f-397dc5d9545e.png">

The reason why we don't see this everytime might be that there might be a race condition with other observers of `UIApplicationDidBecomeActiveNotification`, so that the client ID might be set in time, but not in all cases.

**Comment by tiwoc** [comment]
Hi @aalekz, this is blocking us from upgrading to 2.x. Any chance that a fix could be included in the next release? It looks like it would be sufficient to add `?: [NSNull null]` as in `CASDevice.m:33` in 1.0.10:

https://github.com/castle/castle-ios/blob/f7c6cccae74c8aa9597101664981ea69f8d7d460/Castle/Internal/CASDevice.m#L29-L35

… to `CASContext.m:31` in the 2.x branch:

https://github.com/castle/castle-ios/blob/1066fc026f788ad72197175f2193958e9053414d/Castle/Internal/CASContext.m#L29-L32

**Comment by aalekz** [comment]
Hey @tiwoc!

Sorry for not getting back to you sooner! I'll prepare the fix and do a new release so you can get on with your work.

In version 2.0 we updated the internal fingerprinting used to create the client id, so it should not be possible for it to be null. So this looks like an underlying issue that I'll need to look into. Thank you for bringing it to our attention :)

Please feel free to @mention me in any future issues you might have and I'll help you ASAP.

**Comment by aalekz** [comment]
I just published version [2.0.1](https://github.com/castle/castle-ios/releases/tag/2.0.1) which includes the fix.

**Comment by tiwoc** [comment]
Awesome, thanks!
